tools/codegen/src/generator/languages/rust_lazy_reader/generator.rs

```diff
@@ -0,0 +1,718 @@
+            }
+        };
+        writeln!(output, "{}", q)?;
+
+        for (item_index, item) in self.items().iter().enumerate() {
+            let item_type_name = item.typ().name();
+            let item_type_name = ident_new(&format!("as_{}", item_type_name.to_snake()));
+            let (transformed_name, tc) = get_rust_type_category(item.typ());
+            let convert_code = tc.gen_convert_code();
+            let q = quote! {
+                impl #name {
+                    pub fn #item_type_name(&self) -> Result<#transformed_name, Error> {
+                        let item = self.cursor.union_unpack()?;
+                        if item.item_id != #item_index {
+                            return Err(Error::Header(format!("invalid item type id in union: {}", item.item_id)));
+                        }
+                        let cur = item.cursor.clone();
+                        #convert_code
+                    }
+                }
+            };
+            writeln!(output, "{}", q)?;
+        }
+
+        let verify_items = self.items().iter().enumerate().map(|(_i, item)| {
+            let item_id = item.id();
+            let item_type_name = item.typ().name();
+            let item_type_name = ident_new(&format!("as_{}", item_type_name.to_snake()));
+
+            let typ = item.typ().as_ref();
+
+            let func = match typ {
+                TopDecl::Primitive(v) => {
+                    let fixed_size = v.size() + molecule::NUMBER_SIZE;
+                    quote!(self.cursor.verify_fixed_size(#fixed_size)?;)
+                }
+                TopDecl::FixVec(v) => {
+                    let item_size = v.item_size();
+                    if item_size > 1 {
+                        quote!(
+                            self.#item_type_name()?.verify(compatible)?;
+                        )
+                    } else {
+                        quote!()
+                    }
+                }
+                _ => verify_typ(typ, quote!(self.#item_type_name()?)),
+            };
+            quote!(
+                #item_id => {
+                    #func;
+                    Ok(())
+                }
+            )
+        });
+
+        let q = quote! {
+            impl #name {
+                pub fn verify(&self, compatible: bool) -> Result<(), Error> {
+                    let item_id = self.item_id()?;
+                    match item_id {
+                        #( #verify_items )*
+                        _ => {
+                            Err(Error::UnknownItem(format!("unknow item id: {}", item_id)))
+                        }
+                    }
+                }
+            }
+        };
+        writeln!(output, "{}", q)?;
+        Ok(())
+    }
+}
+
+impl LazyReaderGenerator for ast::Array {
+    fn gen_rust<W: io::Write>(&self, output: &mut W) -> io::Result<()> {
+        let (transformed_name, tc) = get_rust_type_category(self.item().typ());
+        generate_rust_common_array(
```

Comment:
If an array contains a non-trivial type, `get` would also return a `Cursor`, not an actual type. See `Word2` in the test schema for an example. `Word2` contains `Word`, so `get` of `Word2` should return `Word`, not `Cursor`